### ISSUE TITLE
Make results cache TTL configurable per user

### ISSUE THREAD

**Issue by pstibrany** [grafana/mimir#4383]
#### Describe the problem

When user uploads new blocks into Mimir by using backfill feature, currently we recommend flushing the results cache to make new blocks queryable. From Mimir's [documentation](https://grafana.com/docs/mimir/latest/operators-guide/configure/configure-tsdb-block-upload/#the-results-cache-needs-flushing):

> The results-cache needs flushing
> 
> After uploading one or more blocks, the results-cache needs flushing. The reason is that Grafana Mimir caches query results for queries that don’t touch the most recent 10 minutes of data. After uploading blocks however, queries may return different results (because new data was uploaded). Therefore cached results may be wrong, meaning the cache should manually be flushed after uploading blocks.

While this is reasonable behaviour, we would like to give user some guarantee about when backfilled data is fully available for querying, without flushing the entire results cache, because flushing this cache affects ALL users, not only the user with backfilled blocks. This would be especially useful for users who want to use backfill feature regularly, and not just once.

To do that, we would like to lower TTL used by results cache for users who want to use backfill feature. Right now caching TTL is [hardcoded to 7-days](https://github.com/grafana/mimir/blob/0d937fcd431e6648c037a3d21cd3c50a56ab9de5/pkg/frontend/querymiddleware/split_and_cache.go#L35-L38) (and 10m when hitting OOO window), which is a very long time. 

For example reducing cache results TTL for a specific user to 6 hours would give us guarantee that backfilled data is 100% queryable 7-hours after backfill finished (it takes some time for all components to discover newly backfilled blocks, so we're adding extra 1h to accommodate for that).

To this effect, we suggest to add configurable per-tenant TTL for results cache and replace hardcoded 7days (and 10 minutes when hitting OOO).

**Comment by pracucci** [comment]
How would be the end-to-end experience after https://github.com/grafana/mimir/pull/4385 is merged? I'm asking because making the TTL configurable will have an impact on new cache entries, but entries which have already been cached won't change TTL (we're not validating the TTL also when reading back from the cache).

**Comment by pstibrany** [comment]
> How would be the end-to-end experience after #4385 is merged? I'm asking because making the TTL configurable will have an impact on new cache entries, but entries which have already been cached won't change TTL (we're not validating the TTL also when reading back from the cache).

That's a good point. Changing TTL will only affect newly cached results, not previously cached results, so there is still a flush required. This mostly benefits users who want to upload old blocks periodically. Or users who wait before old entries with high TTL expire before doing actual backfill.

**Comment by pracucci** [comment]
> so there is still a flush required

What if we check the TTL of retrieved items too?

**Comment by pracucci** [comment]
> But we would actually need to know the time when the cached entry set in the first place, to check if it is too old. We could add that to our extents. Let me think about it.

Yeah, I was thinking to add the timestamp when the entry was cached as part of the cached item, so that we can check it later when we retrieve it.

**Comment by pstibrany** [comment]
> Yeah, I was thinking to add the timestamp when the entry was cached as part of the cached item, so that we can check it later when we retrieve it.

I will send separate PR adding timestamp to the cached entries, so that we can throw away and rerun queries which already have cached results that should no longer be used due to reduced TTL.

**Comment by pstibrany** [comment]
> I will send separate PR adding timestamp to the cached entries, so that we can throw away and rerun queries which already have cached results that should no longer be used due to reduced TTL.

https://github.com/grafana/mimir/pull/4439

**Comment by pstibrany** [comment]
I think with #4439 merged, we can consider this finished.
